zookeeper/core/component.py

```diff
@@ -104,6 +194,39 @@ def _type_check_and_maybe_cache(instance, field: Field, result: Any) -> None:
         object.__setattr__(instance, field.name, result)
 
 
+def _wrap_init(component_cls: Type) -> None:
+    # If `__init__` is a ComponentInit, we don't want to call that.
+    # This means we subclass another component, in which case we want to call the
+    # original init of the parent class.
+    if isinstance(component_cls.__init__, ComponentInit):
+        original_init = component_cls.__init__.original_init
+    # Otherwise, keep track of any user `__init__`so we can call it after configuration.
```

Comment:
> For example you'd no longer be able to call super().__init__() in the derived class's initializer right?

This is fully possible; I've added a unit test to make sure.

> The name __post_configure__ is better because it makes it clear where it fits in the component creation lifecycle and people don't have the same expectations about that method in the way they do about __init__. The solution I would propose is still not allowing __init__ methods on components but supporting base classes that aren't components that have __init__ methods. ie make [this](https://github.com/larq/zookeeper/blob/main/zookeeper/core/component.py#L720-L724) check only check actual component classes and allow initializer's on non-component base classes.


I thought about this, but it would mean that on components you would do something like
```python
def __post_configure__(self):
    super().__post_configure__()
    # or, if the parent is not decorated with @component:
    super().__init__()
```
and I figured that this would be annoying, since zookeeper components that previously didn't need any kind of `__post_configure__` would all need to add this boilerplate.

Less importantly, we also often have base classes (e.g. `zookeeper.tf.Dataset` and `zookeeper.tf.Preprocessing`) that aren't strictly components (as they're essentially abstract components that are not meant to be ever initialized). If you have such a hierarchy and at some point decide that `YourCustomPreprocessingClass` will no longer be abstract but is instead useful as a component by itself, you would have to change all subclassing components to now call `super().__init__()` instead of `super().__post_configure__()`.


Neither of these are huge drawbacks, but I figured that just always using `__init__` would be cleaner and potentially less confusing.

Reply:
Having said all that, I think you're right and my arguments do not weigh up against the confusion that messing with `__init__` causes (even though zookeeper by definition messes with the `__init__`, but so far it at least prevented you from creating one yourself). I just briefly tested some code and I see no reason why it wouldn't work. It would also make the code a lot simpler, since I basically undo this whole PR and just modify one function instead 🥲 

There is one pitfall: if the user forgets to manually add this `__post_configure__` that calls `super().__init__()`, the instance will still be in an unexpected state after configuration (e.g. it might be a Keras Layer, but it wouldn't function properly). We could throw a warning if this is the case though.

For some additional context: my first idea was to have components not override `__init__` (as they already did), but override `__new__` instead. Unfortunately I could not find a way to instantiate an object of a certain class without having its `__init__` called first.

Reply:
> There is one pitfall: if the user forgets to manually add this `__post_configure__` that calls `super().__init__()`, the instance will still be in an unexpected state after configuration

Could the component call `super().__init__()` automatically if the base class is not a component?

Reply:
In theory yes, but if the `__init__` has any arguments it'll break. So I don't think it'd add too much value.

Reply:
I think you make goods points here. It's complicated. And there are pros and cons to all solutions it seems. My vote is for the simpler change that doesn't mess any more with `__init__` because it's simpler and it doesn't mess with `__init__` :). But the final call is up to you.